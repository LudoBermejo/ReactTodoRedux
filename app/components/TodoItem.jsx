import React from 'react';
import moment from 'moment';

const TodoItem = (props) => {
  const { id, value, completed, createdAt, completedAt } = props;
  const todoClassName = completed ? 'todo-item todo-completed' : 'todo-item';
  function renderCreatedAt(createdAt, completedAt) {
    if (!completedAt) {
      return `Created ${moment.unix(createdAt).format('DD MM YYYY | HH:mm')}`;
    } else {
      return `Completed ${moment.unix(completedAt).format('DD MM YYYY | HH:mm')}`;
    }
  }

  function onChangeComplete(e) {
    e.preventDefault();
    props.onChangeComplete(id);
  }
  return (
    <div
      id={`list_item_${id}`}
      onClick={onChangeComplete}
      className={todoClassName}
    >
      <div>
        <input
          type="checkbox"
          checked={completed}
        />
      </div>
      <div>
        <p>{value}</p>
        <p className="todo__subtext">{renderCreatedAt(createdAt, completedAt)}</p>
      </div>

    </div>
  );
};

TodoItem.propTypes = {
  id: React.PropTypes.string.isRequired,
  value: React.PropTypes.string.isRequired,
  completed: React.PropTypes.bool.isRequired,
  createdAt: React.PropTypes.number.isRequired,
  completedAt: React.PropTypes.number,
  onChangeComplete: React.PropTypes.func.isRequired
};

export default TodoItem;
