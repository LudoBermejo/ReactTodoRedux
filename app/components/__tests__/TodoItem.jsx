import React from 'react';
import renderer from 'react-test-renderer';
import moment from 'moment';

import { shallow, mount } from 'enzyme';
import TodoItem from 'TodoItem';
import uuid from 'node-uuid';

describe('TodoItem', () => {
  it('renders properly', () => {
    const staticValue = {
      id: uuid(),
      value: 'one',
      createdAt: moment().unix(),
      completed: false
    };

    const todoItem = renderer.create(<TodoItem {...staticValue} onChangeComplete={() => {}} />).toJSON();
    expect(todoItem).toMatchSnapshot();
  });

  describe('Todo item show', () => {
    const uuidUnique = uuid();
    const staticValue = {
      id: uuidUnique,
      value: 'one',
      createdAt: moment().unix(),
      completed: false
    };

    const todoItem = shallow(<TodoItem {...staticValue} onChangeComplete={() => {}} />);
    expect(todoItem.find('li').length).toBe(1);
  });

  describe('Todo item change call prop function', () => {
    const uuidUnique = uuid();
    const staticValue = {
      id: uuidUnique,
      value: 'one',
      createdAt: moment().unix(),
      completed: false
    };

    const mock = jest.fn();
    const todoItem = mount(<TodoItem {...staticValue} onChangeComplete={mock} />);

    todoItem.simulate('click');

    expect(mock).toHaveBeenCalledWith(uuidUnique);
  });
});

