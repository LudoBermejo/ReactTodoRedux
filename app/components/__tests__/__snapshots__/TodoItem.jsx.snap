exports[`TodoItem renders properly 1`] = `
<li
  className="todo-item"
  id="list_item_b5bb870f-c100-4c83-8254-5455fadab254"
  onClick={[Function]}>
  <input
    checked={false}
    type="checkbox" />
  <p>
    one
  </p>
  <p>
    Created 14 02 2017 | 12:30
  </p>
</li>
`;
